eureka2-write-server/src/main/java/com/netflix/eureka2/server/transport/tcp/replication/TcpReplicationHandler.java

```diff
@@ -40,35 +48,117 @@
+
+                            }
+
+                            @Override
+                            public void onError(Throwable e) {
+
+                            }
+
+                            @Override
+                            public void onNext(Long aLong) {
+                                logger.info("Evicted {} instances in one round of eviction due to a new receiverReplicationChannel creation", aLong);
+                            }
+                        });
+
+                        return curr;
+                    }
+                })
+                .subscribe(new NoOpSubscriber<ReceiverReplicationChannel>());
     }
 
     @Override
     public Observable<Void> handle(ObservableConnection<Object, Object> connection) {
-        MessageConnection broker = new HeartBeatConnection(
+        final MessageConnection broker = new HeartBeatConnection(
                 new BaseMessageConnection("replicationReceiver", connection, metricFactory.getReplicationReceiverConnectionMetrics()),
                 config.getHeartbeatIntervalMs(), 3,
                 Schedulers.computation()
         );
-        final ServerChannelFactory<ReceiverReplicationChannel> channelFactory =
-                new ReceiverReplicationChannelFactory(registry, broker, SelfIdentityService, evictionQueue, metricFactory);
 
-        return channelFactory.newChannel().asLifecycleObservable();
+        return doHandle(broker).asLifecycleObservable();
     }
+
+    /* visible for testing */ ReceiverReplicationChannel doHandle(MessageConnection connection) {
+        ReceiverReplicationChannel channel =
+                new ReceiverReplicationChannel(connection, selfIdentityService, registry, metricFactory.getReplicationChannelMetrics());
+
+        channelSubject.onNext(channel);
+
+        return channel;
+    }
+
+    private Observable<Long> setUpEviction(final ReceiverReplicationChannel prev, final ReceiverReplicationChannel curr) {
+        // once a new channel is available, wait for the first bufferEnd to be emitted. Once it is emitted,
+        // return a reference to the previous channel.
+
+        return curr.getStreamStateNotifications()
+                .filter(new Func1<ChangeNotification<InstanceInfo>, Boolean>() {
+                    @Override
+                    public Boolean call(ChangeNotification<InstanceInfo> notification) {
+                        if (notification instanceof StreamStateNotification) {
+                            StreamStateNotification<InstanceInfo> n = (StreamStateNotification<InstanceInfo>) notification;
+                            if (n.getBufferState() == StreamStateNotification.BufferState.BufferEnd) {
+                                return true;
+                            }
+                        }
+                        return false;
+                    }
+                })
+                .take(1)
+                .map(new Func1<ChangeNotification<InstanceInfo>, ReceiverReplicationChannel>() {
```

Comment:
It's not. In the interest channel version (from which this was copied) it was needed.